chia/full_node/mempool_manager.py

```diff
@@ -142,35 +143,82 @@ async def create_bundle_from_mempool(
 
         cost_sum = 0  # Checks that total cost does not exceed block maximum
         fee_sum = 0  # Checks that total fees don't exceed 64 bits
-        spend_bundles: List[SpendBundle] = []
-        removals = []
-        additions = []
+        processed_spend_bundles = 0
+        removals: List[Coin] = []
+        additions: Set[Coin] = set()
+        # This is a map of coin ID to a pair of the coin spend solution hash and its isolated cost
+        dedup_coin_spends: Dict[bytes32, Tuple[bytes32, uint64]] = {}
+        coin_spends: List[CoinSpend] = []
+        sigs: List[G2Element] = []
         broke_from_inner_loop = False
         log.info(f"Starting to make block, max cost: {self.constants.MAX_BLOCK_COST_CLVM}")
         for dic in reversed(self.mempool.sorted_spends.values()):
             if broke_from_inner_loop:
                 break
             for item in dic.values():
-                log.info(f"Cumulative cost: {cost_sum}, fee per cost: {item.fee / item.cost}")
-                if (
-                    item.cost + cost_sum <= self.limit_factor * self.constants.MAX_BLOCK_COST_CLVM
-                    and item.fee + fee_sum <= self.constants.MAX_COIN_AMOUNT
-                ):
-                    spend_bundles.append(item.spend_bundle)
-                    cost_sum += item.cost
-                    fee_sum += item.fee
-                    removals.extend(item.removals)
-                    additions.extend(item.additions)
-                else:
-                    broke_from_inner_loop = True
-                    break
-        if len(spend_bundles) > 0:
+                consider_this_item = True
+                cost_saving = 0
+                spends_to_dedup: List[CoinSpend] = []
+                spends_to_run_for_cost: List[Tuple[bytes32, CoinSpend]] = []
```

Comment:
I really don't think we should compute the isolated cost for a puzzle+solution until we need to. so, the first time we find a duplicate, that's when we should compute this cost. I imagine removing this list and making the cost field optional in `dedup_coin_spends`. The first time you need the cost, compute it then.

Also, computing the cost should be its own function that you can unit test independently.